SocketTimeoutException from fetching number of images deleted

**Summary:** 

I logged in to Misaochan, went to Nearby, and the app crashed for me while I was searching for an image to upload to Nearby. Stack trace seems to point to getRevertRespObjectSingle() in ApacheHttpClientMediaWikiApi

**Steps to reproduce:** 

Unsure of how to reproduce, it is not consistent. However, we should handle this exception more gracefully. I think catching it and outputting an error message is sufficient since that method is not essential (it's just used to get user revert rate). [Edit: After looking into it further, this is NOT true - this function is called whenever the user loads ContributionsActivity, and it is nonNull.]

**Add System logs:**

```
09-03 19:16:22.913 27648-27648/fr.free.nrw.commons W/System.err: Caused by: java.net.SocketTimeoutException: timeout
        at okhttp3.internal.http2.Http2Stream$StreamTimeout.newTimeoutException(Http2Stream.java:593)
        at okhttp3.internal.http2.Http2Stream$StreamTimeout.exitAndThrowIfTimedOut(Http2Stream.java:601)
09-03 19:16:22.914 27648-27648/fr.free.nrw.commons W/System.err:     at okhttp3.internal.http2.Http2Stream.takeResponseHeaders(Http2Stream.java:146)
        at okhttp3.internal.http2.Http2Codec.readResponseHeaders(Http2Codec.java:125)
        at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:88)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
09-03 19:16:22.915 27648-27648/fr.free.nrw.commons W/System.err:     at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:45)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
09-03 19:16:22.916 27648-27648/fr.free.nrw.commons W/System.err:     at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
        at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
09-03 19:16:22.917 27648-27648/fr.free.nrw.commons W/System.err:     at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
        at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:126)
09-03 19:16:22.918 27648-27648/fr.free.nrw.commons W/System.err:     at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
        at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:200)
        at okhttp3.RealCall.execute(RealCall.java:77)
09-03 19:16:22.919 27648-27648/fr.free.nrw.commons W/System.err:     at fr.free.nrw.commons.mwapi.ApacheHttpClientMediaWikiApi.lambda$getRevertRespObjectSingle$5$ApacheHttpClientMediaWikiApi(ApacheHttpClientMediaWikiApi.java:1018)
        at fr.free.nrw.commons.mwapi.ApacheHttpClientMediaWikiApi$$Lambda$9.call(Unknown Source:4)
        at io.reactivex.internal.operators.single.SingleFromCallable.subscribeActual(SingleFromCallable.java:35)
09-03 19:16:22.920 27648-27648/fr.free.nrw.commons W/System.err:     at io.reactivex.Single.subscribe(Single.java:2700)
        at com.tspoon.traceur.SingleOnAssembly.subscribeActual(SingleOnAssembly.java:43)
        at io.reactivex.Single.subscribe(Single.java:2700)
        at io.reactivex.internal.operators.single.SingleSubscribeOn$SubscribeOnObserver.run(SingleSubscribeOn.java:89)
09-03 19:16:22.921 27648-27648/fr.free.nrw.commons W/System.err:     at io.reactivex.Scheduler$DisposeTask.run(Scheduler.java:452)
        at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:61)
09-03 19:16:22.922 27648-27648/fr.free.nrw.commons W/System.err:     at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:52)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
09-03 19:16:22.923 27648-27648/fr.free.nrw.commons W/System.err:     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
        at java.lang.Thread.run(Thread.java:764)
09-03 19:16:22.924 27648-27648/fr.free.nrw.commons W/System.err: Caused by: com.tspoon.traceur.TraceurException: Debug Exception generated at call site
        at dalvik.system.VMStack.getThreadStackTrace(Native Method)
        at java.lang.Thread.getStackTrace(Thread.java:1536)
        at io.reactivex.Single.fromCallable(Single.java:448)
09-03 19:16:22.925 27648-27648/fr.free.nrw.commons W/System.err:     at fr.free.nrw.commons.mwapi.ApacheHttpClientMediaWikiApi.getRevertRespObjectSingle(ApacheHttpClientMediaWikiApi.java:1005)
        at fr.free.nrw.commons.quiz.QuizChecker.setRevertCount(QuizChecker.java:92)
        at fr.free.nrw.commons.quiz.QuizChecker.<init>(QuizChecker.java:55)
09-03 19:16:22.926 27648-27648/fr.free.nrw.commons W/System.err:     at fr.free.nrw.commons.contributions.ContributionsActivity.onCreate(ContributionsActivity.java:152)
        at android.app.Activity.performCreate(Activity.java:7174)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1220)
09-03 19:16:22.927 27648-27648/fr.free.nrw.commons W/System.err:     at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2910)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3032)
        at android.app.ActivityThread.-wrap11(Unknown Source:0)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1696)
09-03 19:16:22.928 27648-27648/fr.free.nrw.commons W/System.err:     at android.os.Handler.dispatchMessage(Handler.java:105)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6944)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:327)
09-03 19:16:22.929 27648-27648/fr.free.nrw.commons W/System.err:     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1374)
09-03 19:16:22.930 27648-27648/fr.free.nrw.commons W/System.err: Caused by: com.tspoon.traceur.TraceurException: Debug Exception generated at call site
        at dalvik.system.VMStack.getThreadStackTrace(Native Method)
        at java.lang.Thread.getStackTrace(Thread.java:1536)
09-03 19:16:22.931 27648-27648/fr.free.nrw.commons W/System.err:     at io.reactivex.Single.subscribeOn(Single.java:2768)
        at fr.free.nrw.commons.quiz.QuizChecker.setRevertCount(QuizChecker.java:93)
09-03 19:16:22.932 27648-27648/fr.free.nrw.commons W/System.err: 	... 14 more
    Caused by: com.tspoon.traceur.TraceurException: Debug Exception generated at call site
09-03 19:16:22.933 27648-27648/fr.free.nrw.commons W/System.err:     at dalvik.system.VMStack.getThreadStackTrace(Native Method)
        at java.lang.Thread.getStackTrace(Thread.java:1536)
        at io.reactivex.Single.observeOn(Single.java:2296)
        at fr.free.nrw.commons.quiz.QuizChecker.setRevertCount(QuizChecker.java:94)
09-03 19:16:22.934 27648-27648/fr.free.nrw.commons W/System.err: 	... 14 more
09-03 19:16:22.946 3646-8649/? W/ActivityManager: crash : fr.free.nrw.commons,0
09-03 19:16:22.948 3646-8649/? W/ActivityManager:   Force finishing activity fr.free.nrw.commons/.nearby.NearbyActivity
```

**Device and Android version:** 

Android 8.0 Samsung Galaxy s7
 
 **Commons app version:** 

https://github.com/commons-app/apps-android-commons/pull/1875 which is based on `2.8-release`, but the error does not appear to be relevant to that PR at all, just coincidence

**Would you like to work on the issue?**

Pref not

--- a/app/src/main/java/fr/free/nrw/commons/achievements/Achievements.java
+++ b/app/src/main/java/fr/free/nrw/commons/achievements/Achievements.java
@@ -1,7 +1,5 @@
 package fr.free.nrw.commons.achievements;
 
-import android.util.Log;
-
 /**
  * represnts Achievements class ans stores all the parameters
  */
@@ -42,6 +40,23 @@ public class Achievements {
         this.featuredImages = featuredImages;
         this.imagesUploaded = imagesUploaded;
         this.revertCount = revertCount;
+    }
+
+    /**
+     * Get Achievements object from FeedbackResponse
+     *
+     * @param response
+     * @return
+     */
+    public static Achievements from(FeedbackResponse response) {
+        return new Achievements(response.getUniqueUsedImages(),
+                response.getArticlesUsingImages(),
+                response.getThanksReceived(),
+                response.getImagesEditedBySomeoneElse(),
+                response.getFeaturedImages().getQualityImages()
+                        + response.getFeaturedImages().getFeaturedPicturesOnWikimediaCommons(),
+                0,
+                response.getDeletedUploads());
     }
 
     /**

--- a/app/src/main/java/fr/free/nrw/commons/achievements/AchievementsActivity.java
+++ b/app/src/main/java/fr/free/nrw/commons/achievements/AchievementsActivity.java
@@ -2,16 +2,9 @@ package fr.free.nrw.commons.achievements;
 
 import android.accounts.Account;
 import android.annotation.SuppressLint;
-import android.annotation.TargetApi;
 import android.content.Context;
-import android.content.DialogInterface;
 import android.content.Intent;
 import android.graphics.Bitmap;
-import android.graphics.BitmapFactory;
-import android.graphics.Canvas;
-import android.graphics.Color;
-import android.graphics.Paint;
-import android.graphics.Rect;
 import android.graphics.drawable.BitmapDrawable;
 import android.graphics.drawable.Drawable;
 import android.net.Uri;
@@ -20,7 +13,6 @@ import android.support.v4.content.res.ResourcesCompat;
 import android.support.v7.app.AlertDialog;
 import android.support.v7.widget.Toolbar;
 import android.util.DisplayMetrics;
-import android.util.Log;
 import android.view.ContextThemeWrapper;
 import android.view.LayoutInflater;
 import android.view.Menu;
@@ -34,13 +26,10 @@ import android.widget.TextView;
 
 import com.dinuscxj.progressbar.CircleProgressBar;
 
-import org.json.JSONException;
-import org.json.JSONObject;
-
 import java.io.File;
 import java.io.FileOutputStream;
 import java.io.IOException;
-import java.util.Optional;
+import java.util.Objects;
 
 import javax.inject.Inject;
 
@@ -141,7 +130,6 @@ public class AchievementsActivity extends NavigationBaseActivity {
         hideLayouts();
         setAchievements();
         setUploadCount();
-        setRevertCount();
         initDrawer();
     }
 
@@ -201,48 +189,24 @@ public class AchievementsActivity extends NavigationBaseActivity {
      * which then calls parseJson when results are fetched
      */
     private void setAchievements() {
+
         if(checkAccount()) {
             compositeDisposable.add(mediaWikiApi
-                    .getAchievements(sessionManager.getCurrentAccount().name)
+                    .getAchievements(Objects.requireNonNull(sessionManager.getCurrentAccount()).name)
                     .subscribeOn(Schedulers.io())
                     .observeOn(AndroidSchedulers.mainThread())
                     .subscribe(
-                            jsonObject -> parseJson(jsonObject),
+                            response -> {
+                                if (response != null) {
+                                    achievements = Achievements.from(response);
+                                    isRevertFetched = true;
+                                    isStatisticsFetched = true;
+                                }
+                                hideProgressBar();
+                            },
                             t -> Timber.e(t, "Fetching achievements statisticss failed")
                     ));
         }
-    }
-
-    /**
-     * To call the API to get reverts count in form of JSONObject
-     *
-     */
-
-    private void setRevertCount(){
-        if(checkAccount()) {
-            compositeDisposable.add(mediaWikiApi
-                    .getRevertRespObjectSingle(sessionManager.getCurrentAccount().name)
-                    .subscribeOn(Schedulers.io())
-                    .observeOn(AndroidSchedulers.mainThread())
-                    .subscribe(
-                            object -> parseJsonRevertCount(object),
-                            t -> Timber.e(t, "Fetching revert count failed")
-                    ));
-        }
-    }
-
-    /**
-     * used to set number of deleted images
-     * @param object
-     */
-    private void parseJsonRevertCount(JSONObject object){
-        try {
-            achievements.setRevertCount(object.getInt("deletedUploads"));
-        } catch (JSONException e) {
-            Timber.d( e, e.getMessage());
-        }
-        isRevertFetched = true;
-        hideProgressBar();
     }
 
     /**
@@ -251,11 +215,11 @@ public class AchievementsActivity extends NavigationBaseActivity {
     private void setUploadCount() {
         if(checkAccount()) {
             compositeDisposable.add(mediaWikiApi
-                    .getUploadCount(sessionManager.getCurrentAccount().name)
+                    .getUploadCount(Objects.requireNonNull(sessionManager.getCurrentAccount()).name)
                     .subscribeOn(Schedulers.io())
                     .observeOn(AndroidSchedulers.mainThread())
                     .subscribe(
-                            uploadCount -> setAchievementsUploadCount(uploadCount),
+                            this::setAchievementsUploadCount,
                             t -> Timber.e(t, "Fetching upload count failed")
                     ));
         }
@@ -294,32 +258,11 @@ public class AchievementsActivity extends NavigationBaseActivity {
     }
 
     /**
-     * used to parse the JSONObject containing results
-     * @param object
-     */
-    private void parseJson(JSONObject object) {
-        try {
-            achievements.setUniqueUsedImages(object.getInt("uniqueUsedImages"));
-            achievements.setArticlesUsingImages(object.getInt("articlesUsingImages"));
-            achievements.setThanksReceived(object.getInt("thanksReceived"));
-            achievements.setImagesEditedBySomeoneElse(object.getInt("imagesEditedBySomeoneElse"));
-            JSONObject featuredImages = object.getJSONObject("featuredImages");
-            achievements.setFeaturedImages
-                    (featuredImages.getInt("Quality_images") +
-                            featuredImages.getInt("Featured_pictures_on_Wikimedia_Commons"));
-        } catch (JSONException e) {
-            e.printStackTrace();
-        }
-        isStatisticsFetched = true;
-        hideProgressBar();
-    }
-
-    /**
      * Used the inflate the fetched statistics of the images uploaded by user
      * and assign badge and level
      * @param achievements
      */
-    private void inflateAchievements(Achievements achievements ){
+    private void inflateAchievements(Achievements achievements) {
         thanksReceived.setText(Integer.toString(achievements.getThanksReceived()));
         imagesUsedByWikiProgessbar.setProgress
                 (100*achievements.getUniqueUsedImages()/levelInfo.getMaxUniqueImages() );
@@ -393,17 +336,8 @@ public class AchievementsActivity extends NavigationBaseActivity {
         TextView shareMessage = (TextView) view.findViewById(R.id.alert_text);
         shareMessage.setText(R.string.achievements_share_message);
         alertadd.setView(view);
-        alertadd.setPositiveButton("Proceed", new DialogInterface.OnClickListener() {
-            public void onClick(DialogInterface dialog, int which) {
-                shareScreen(screenshot);
-            }
-        });
-        alertadd.setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
-            @Override
-            public void onClick(DialogInterface dialog, int which) {
-                dialog.cancel();
-            }
-        });
+        alertadd.setPositiveButton("Proceed", (dialog, which) -> shareScreen(screenshot));
+        alertadd.setNegativeButton("Cancel", (dialog, which) -> dialog.cancel());
         alertadd.show();
     }
 

--- a/app/src/main/java/fr/free/nrw/commons/achievements/FeaturedImages.java
+++ b/app/src/main/java/fr/free/nrw/commons/achievements/FeaturedImages.java
@@ -1,0 +1,25 @@
+package fr.free.nrw.commons.achievements;
+
+import com.google.gson.annotations.SerializedName;
+
+public class FeaturedImages {
+
+    @SerializedName("Quality_images")
+    private final int qualityImages;
+
+    @SerializedName("Featured_pictures_on_Wikimedia_Commons")
+    private final int featuredPicturesOnWikimediaCommons;
+
+    public FeaturedImages(int qualityImages, int featuredPicturesOnWikimediaCommons) {
+        this.qualityImages = qualityImages;
+        this.featuredPicturesOnWikimediaCommons = featuredPicturesOnWikimediaCommons;
+    }
+
+    public int getQualityImages() {
+        return qualityImages;
+    }
+
+    public int getFeaturedPicturesOnWikimediaCommons() {
+        return featuredPicturesOnWikimediaCommons;
+    }
+}

--- a/app/src/main/java/fr/free/nrw/commons/achievements/FeedbackResponse.java
+++ b/app/src/main/java/fr/free/nrw/commons/achievements/FeedbackResponse.java
@@ -1,0 +1,64 @@
+package fr.free.nrw.commons.achievements;
+
+public class FeedbackResponse {
+
+    private final String status;
+    private final int uniqueUsedImages;
+    private final int articlesUsingImages;
+    private final int deletedUploads;
+    private final FeaturedImages featuredImages;
+    private final int thanksReceived;
+    private final String user;
+    private final int imagesEditedBySomeoneElse;
+
+
+    public FeedbackResponse(String status,
+                            int uniqueUsedImages,
+                            int articlesUsingImages,
+                            int deletedUploads,
+                            FeaturedImages featuredImages,
+                            int thanksReceived,
+                            String user,
+                            int imagesEditedBySomeoneElse) {
+        this.status = status;
+        this.uniqueUsedImages = uniqueUsedImages;
+        this.articlesUsingImages = articlesUsingImages;
+        this.deletedUploads = deletedUploads;
+        this.featuredImages = featuredImages;
+        this.thanksReceived = thanksReceived;
+        this.user = user;
+        this.imagesEditedBySomeoneElse = imagesEditedBySomeoneElse;
+    }
+
+    public String getStatus() {
+        return status;
+    }
+
+    public int getUniqueUsedImages() {
+        return uniqueUsedImages;
+    }
+
+    public int getArticlesUsingImages() {
+        return articlesUsingImages;
+    }
+
+    public int getDeletedUploads() {
+        return deletedUploads;
+    }
+
+    public FeaturedImages getFeaturedImages() {
+        return featuredImages;
+    }
+
+    public int getThanksReceived() {
+        return thanksReceived;
+    }
+
+    public String getUser() {
+        return user;
+    }
+
+    public int getImagesEditedBySomeoneElse() {
+        return imagesEditedBySomeoneElse;
+    }
+}

--- a/app/src/main/java/fr/free/nrw/commons/di/CommonsApplicationModule.java
+++ b/app/src/main/java/fr/free/nrw/commons/di/CommonsApplicationModule.java
@@ -11,7 +11,6 @@ import javax.inject.Singleton;
 
 import dagger.Module;
 import dagger.Provides;
-
 import fr.free.nrw.commons.BuildConfig;
 import fr.free.nrw.commons.auth.AccountUtil;
 import fr.free.nrw.commons.auth.SessionManager;
@@ -22,6 +21,7 @@ import fr.free.nrw.commons.nearby.NearbyPlaces;
 import fr.free.nrw.commons.upload.UploadController;
 import fr.free.nrw.commons.wikidata.WikidataEditListener;
 import fr.free.nrw.commons.wikidata.WikidataEditListenerImpl;
+import okhttp3.OkHttpClient;
 
 import static android.content.Context.MODE_PRIVATE;
 import static fr.free.nrw.commons.explore.recentsearches.RecentSearchesContentProvider.RECENT_SEARCH_AUTHORITY;

--- a/app/src/main/java/fr/free/nrw/commons/di/NetworkingModule.java
+++ b/app/src/main/java/fr/free/nrw/commons/di/NetworkingModule.java
@@ -7,6 +7,8 @@ import android.support.annotation.NonNull;
 import com.google.gson.Gson;
 import com.google.gson.GsonBuilder;
 
+import java.io.File;
+
 import javax.inject.Named;
 import javax.inject.Singleton;
 
@@ -15,6 +17,7 @@ import dagger.Provides;
 import fr.free.nrw.commons.BuildConfig;
 import fr.free.nrw.commons.mwapi.ApacheHttpClientMediaWikiApi;
 import fr.free.nrw.commons.mwapi.MediaWikiApi;
+import okhttp3.Cache;
 import okhttp3.HttpUrl;
 import okhttp3.OkHttpClient;
 
@@ -25,8 +28,11 @@ public class NetworkingModule {
 
     @Provides
     @Singleton
-    public OkHttpClient provideOkHttpClient() {
-        return new OkHttpClient.Builder().build();
+    public OkHttpClient provideOkHttpClient(Context context) {
+        File dir = new File(context.getCacheDir(), "okHttpCache");
+        return new OkHttpClient.Builder()
+                .cache(new Cache(dir, OK_HTTP_CACHE_SIZE))
+                .build();
     }
 
     @Provides
@@ -34,8 +40,9 @@ public class NetworkingModule {
     public MediaWikiApi provideMediaWikiApi(Context context,
                                             @Named("default_preferences") SharedPreferences defaultPreferences,
                                             @Named("category_prefs") SharedPreferences categoryPrefs,
-                                            Gson gson) {
-        return new ApacheHttpClientMediaWikiApi(context, BuildConfig.WIKIMEDIA_API_HOST, BuildConfig.WIKIDATA_API_HOST, defaultPreferences, categoryPrefs, gson);
+                                            Gson gson,
+                                            OkHttpClient okHttpClient) {
+        return new ApacheHttpClientMediaWikiApi(context, BuildConfig.WIKIMEDIA_API_HOST, BuildConfig.WIKIDATA_API_HOST, defaultPreferences, categoryPrefs, gson, okHttpClient);
     }
 
     @Provides

--- a/app/src/main/java/fr/free/nrw/commons/mwapi/ApacheHttpClientMediaWikiApi.java
+++ b/app/src/main/java/fr/free/nrw/commons/mwapi/ApacheHttpClientMediaWikiApi.java
@@ -24,7 +24,6 @@ import org.apache.http.impl.conn.tsccm.ThreadSafeClientConnManager;
 import org.apache.http.params.BasicHttpParams;
 import org.apache.http.params.CoreProtocolPNames;
 import org.apache.http.util.EntityUtils;
-import org.json.JSONObject;
 import org.w3c.dom.Element;
 import org.w3c.dom.Node;
 import org.w3c.dom.NodeList;
@@ -45,6 +44,7 @@ import java.util.concurrent.Callable;
 import fr.free.nrw.commons.BuildConfig;
 import fr.free.nrw.commons.Media;
 import fr.free.nrw.commons.PageTitle;
+import fr.free.nrw.commons.achievements.FeedbackResponse;
 import fr.free.nrw.commons.auth.AccountUtil;
 import fr.free.nrw.commons.category.CategoryImageUtils;
 import fr.free.nrw.commons.category.QueryContinue;
@@ -76,14 +76,17 @@ public class ApacheHttpClientMediaWikiApi implements MediaWikiApi {
     private SharedPreferences defaultPreferences;
     private SharedPreferences categoryPreferences;
     private Gson gson;
+    private final OkHttpClient okHttpClient;
 
     public ApacheHttpClientMediaWikiApi(Context context,
                                         String apiURL,
                                         String wikidatApiURL,
                                         SharedPreferences defaultPreferences,
                                         SharedPreferences categoryPreferences,
-                                        Gson gson) {
+                                        Gson gson,
+                                        OkHttpClient okHttpClient) {
         this.context = context;
+        this.okHttpClient = okHttpClient;
         BasicHttpParams params = new BasicHttpParams();
         SchemeRegistry schemeRegistry = new SchemeRegistry();
         schemeRegistry.register(new Scheme("http", PlainSocketFactory.getSocketFactory(), 80));
@@ -966,9 +969,8 @@ public class ApacheHttpClientMediaWikiApi implements MediaWikiApi {
      * @param userName
      * @return
      */
-    @NonNull
     @Override
-    public Single<JSONObject> getAchievements(String userName) {
+    public Single<FeedbackResponse> getAchievements(String userName) {
         final String fetchAchievementUrlTemplate =
                 wikiMediaToolforgeUrl + "urbanecmbot/commonsmisc/feedback.py";
         return Single.fromCallable(() -> {
@@ -982,44 +984,14 @@ public class ApacheHttpClientMediaWikiApi implements MediaWikiApi {
             Request request = new Request.Builder()
                     .url(urlBuilder.toString())
                     .build();
-            OkHttpClient client = new OkHttpClient();
-            Response response = client.newCall(request).execute();
-            String jsonData = response.body().string();
-            JSONObject jsonObject = new JSONObject(jsonData);
-            return jsonObject;
+            Response response = okHttpClient.newCall(request).execute();
+            if (response.body() != null && response.isSuccessful()) {
+                String json = response.body().string();
+                return gson.fromJson(json, FeedbackResponse.class);
+            }
+            return null;
         });
 
-    }
-
-    /**
-     * This takes userName as input, which is then used to fetch the no of images deleted
-     * using OkHttp and JavaRx. This function return JSONObject
-     * @param userName
-     * @return
-     */
-    @NonNull
-    @Override
-    public Single<JSONObject> getRevertRespObjectSingle(String userName){
-        final String fetchRevertCountUrlTemplate =
-                wikiMediaToolforgeUrl + "urbanecmbot/commonsmisc/feedback.py";
-        return Single.fromCallable(() -> {
-            String url = String.format(
-                    Locale.ENGLISH,
-                    fetchRevertCountUrlTemplate,
-                    new PageTitle(userName).getText());
-            HttpUrl.Builder urlBuilder = HttpUrl.parse(url).newBuilder();
-            urlBuilder.addQueryParameter("user", userName);
-            urlBuilder.addQueryParameter("fetch","deletedUploads");
-            Log.i("url", urlBuilder.toString());
-            Request request = new Request.Builder()
-                    .url(urlBuilder.toString())
-                    .build();
-            OkHttpClient client = new OkHttpClient();
-            Response response = client.newCall(request).execute();
-            String jsonData = response.body().string();
-            JSONObject jsonRevertObject = new JSONObject(jsonData);
-            return jsonRevertObject;
-        });
     }
 
     private Date parseMWDate(String mwDate) {

--- a/app/src/main/java/fr/free/nrw/commons/mwapi/ApacheHttpClientMediaWikiApi.java
+++ b/app/src/main/java/fr/free/nrw/commons/mwapi/ApacheHttpClientMediaWikiApi.java
@@ -966,7 +966,7 @@ public class ApacheHttpClientMediaWikiApi implements MediaWikiApi {
     /**
      * This takes userName as input, which is then used to fetch the feedback/achievements
      * statistics using OkHttp and JavaRx. This function return JSONObject
-     * @param userName
+     * @param userName MediaWiki user name
      * @return
      */
     @Override
@@ -985,8 +985,11 @@ public class ApacheHttpClientMediaWikiApi implements MediaWikiApi {
                     .url(urlBuilder.toString())
                     .build();
             Response response = okHttpClient.newCall(request).execute();
-            if (response.body() != null && response.isSuccessful()) {
+            if (response != null && response.body() != null && response.isSuccessful()) {
                 String json = response.body().string();
+                if (json == null) {
+                    return null;
+                }
                 return gson.fromJson(json, FeedbackResponse.class);
             }
             return null;

--- a/app/src/main/java/fr/free/nrw/commons/mwapi/MediaWikiApi.java
+++ b/app/src/main/java/fr/free/nrw/commons/mwapi/MediaWikiApi.java
@@ -99,7 +99,6 @@ public interface MediaWikiApi {
 
     boolean isUserBlockedFromCommons();
 
-    @NonNull
     Single<FeedbackResponse> getAchievements(String userName);
 
     interface ProgressListener {

--- a/app/src/main/java/fr/free/nrw/commons/mwapi/MediaWikiApi.java
+++ b/app/src/main/java/fr/free/nrw/commons/mwapi/MediaWikiApi.java
@@ -4,13 +4,12 @@ import android.net.Uri;
 import android.support.annotation.NonNull;
 import android.support.annotation.Nullable;
 
-import org.json.JSONObject;
-
 import java.io.IOException;
 import java.io.InputStream;
 import java.util.List;
 
 import fr.free.nrw.commons.Media;
+import fr.free.nrw.commons.achievements.FeedbackResponse;
 import fr.free.nrw.commons.notification.Notification;
 import io.reactivex.Observable;
 import io.reactivex.Single;
@@ -98,13 +97,10 @@ public interface MediaWikiApi {
     @NonNull
     Single<Integer> getUploadCount(String userName);
 
-    @NonNull
-    Single<JSONObject> getRevertRespObjectSingle(String userName);
-
     boolean isUserBlockedFromCommons();
 
     @NonNull
-    Single<JSONObject> getAchievements(String userName);
+    Single<FeedbackResponse> getAchievements(String userName);
 
     interface ProgressListener {
         void onProgress(long transferred, long total);

--- a/app/src/main/java/fr/free/nrw/commons/quiz/QuizChecker.java
+++ b/app/src/main/java/fr/free/nrw/commons/quiz/QuizChecker.java
@@ -5,7 +5,6 @@ import android.content.DialogInterface;
 import android.content.Intent;
 import android.content.SharedPreferences;
 import android.support.v7.app.AlertDialog.Builder;
-import android.util.Log;
 
 import fr.free.nrw.commons.R;
 import fr.free.nrw.commons.WelcomeActivity;
@@ -89,12 +88,16 @@ public class QuizChecker {
      */
     private void setRevertCount() {
             compositeDisposable.add(mediaWikiApi
-                    .getRevertRespObjectSingle(userName)
+                    .getAchievements(userName)
                     .subscribeOn(Schedulers.io())
                     .observeOn(AndroidSchedulers.mainThread())
                     .subscribe(
-                            object -> setRevertParameter(object.getInt("deletedUploads"))
-                    ));
+                            response -> {
+                                if (response != null) {
+                                    setRevertParameter(response.getDeletedUploads());
+                                }
+                            })
+            );
     }
 
     /**

--- a/app/src/main/java/fr/free/nrw/commons/quiz/QuizChecker.java
+++ b/app/src/main/java/fr/free/nrw/commons/quiz/QuizChecker.java
@@ -1,7 +1,6 @@
 package fr.free.nrw.commons.quiz;
 
 import android.content.Context;
-import android.content.DialogInterface;
 import android.content.Intent;
 import android.content.SharedPreferences;
 import android.support.v7.app.AlertDialog.Builder;
@@ -33,16 +32,16 @@ public class QuizChecker {
     private SharedPreferences revertPref;
     private SharedPreferences countPref;
 
-    private final int UPLOAD_COUNT_THRESHOLD = 5;
-    private final String REVERT_PERCENTAGE_FOR_MESSAGE = "50%";
+    private static final int UPLOAD_COUNT_THRESHOLD = 5;
+    private static final String REVERT_PERCENTAGE_FOR_MESSAGE = "50%";
     private final String REVERT_SHARED_PREFERENCE = "revertCount";
     private final String UPLOAD_SHARED_PREFERENCE = "uploadCount";
 
     /**
      * constructor to set the parameters for quiz
-     * @param context
-     * @param userName
-     * @param mediaWikiApi
+     * @param context context
+     * @param userName Commons user name
+     * @param mediaWikiApi instance of MediaWikiApi
      */
     public QuizChecker(Context context, String userName, MediaWikiApi mediaWikiApi) {
         this.context = context;
@@ -62,8 +61,7 @@ public class QuizChecker {
                     .getUploadCount(userName)
                     .subscribeOn(Schedulers.io())
                     .observeOn(AndroidSchedulers.mainThread())
-                    .subscribe(
-                            uploadCount -> setTotalUploadCount(uploadCount),
+                    .subscribe(this::setTotalUploadCount,
                             t -> Timber.e(t, "Fetching upload count failed")
                     ));
     }
@@ -71,7 +69,7 @@ public class QuizChecker {
     /**
      * set the sub Title of Contibutions Activity and
      * call function to check for quiz
-     * @param uploadCount
+     * @param uploadCount user's upload count
      */
     private void setTotalUploadCount(int uploadCount) {
         totalUploadCount = uploadCount - countPref.getInt(UPLOAD_SHARED_PREFERENCE,0);
@@ -102,7 +100,7 @@ public class QuizChecker {
 
     /**
      * to calculate the number of images reverted after previous quiz
-     * @param revertCountFetched
+     * @param revertCountFetched count of deleted uploads
      */
     private void setRevertParameter(int revertCountFetched) {
         revertCount = revertCountFetched - revertPref.getInt(REVERT_SHARED_PREFERENCE,0);
@@ -133,30 +131,22 @@ public class QuizChecker {
     /**
      * Alert which prompts to quiz
      */
-    public void callQuiz() {
+    private void callQuiz() {
         Builder alert = new Builder(context);
         alert.setTitle(context.getResources().getString(R.string.quiz));
         alert.setMessage(context.getResources().getString(R.string.quiz_alert_message,
                 REVERT_PERCENTAGE_FOR_MESSAGE));
-        alert.setPositiveButton("Proceed", new DialogInterface.OnClickListener() {
-            @Override
-            public void onClick(DialogInterface dialog, int which) {
-               int newRevetSharedPrefs = revertCount+ revertPref.getInt(REVERT_SHARED_PREFERENCE,0);
-                revertPref.edit().putInt(REVERT_SHARED_PREFERENCE, newRevetSharedPrefs).apply();
-                int newUploadCount = totalUploadCount + countPref.getInt(UPLOAD_SHARED_PREFERENCE,0);
-                countPref.edit().putInt(UPLOAD_SHARED_PREFERENCE,newUploadCount).apply();
-                Intent i = new Intent(context, WelcomeActivity.class);
-                i.putExtra("isQuiz", true);
-                dialog.dismiss();
-                context.startActivity(i);
-            }
+        alert.setPositiveButton("Proceed", (dialog, which) -> {
+            int newRevetSharedPrefs = revertCount + revertPref.getInt(REVERT_SHARED_PREFERENCE, 0);
+            revertPref.edit().putInt(REVERT_SHARED_PREFERENCE, newRevetSharedPrefs).apply();
+            int newUploadCount = totalUploadCount + countPref.getInt(UPLOAD_SHARED_PREFERENCE, 0);
+            countPref.edit().putInt(UPLOAD_SHARED_PREFERENCE, newUploadCount).apply();
+            Intent i = new Intent(context, WelcomeActivity.class);
+            i.putExtra("isQuiz", true);
+            dialog.dismiss();
+            context.startActivity(i);
         });
-        alert.setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
-            @Override
-            public void onClick(DialogInterface dialogInterface, int i) {
-                dialogInterface.cancel();
-            }
-        });
+        alert.setNegativeButton("Cancel", (dialogInterface, i) -> dialogInterface.cancel());
         android.support.v7.app.AlertDialog dialog = alert.create();
         dialog.show();
     }

--- a/app/src/main/java/fr/free/nrw/commons/quiz/QuizChecker.java
+++ b/app/src/main/java/fr/free/nrw/commons/quiz/QuizChecker.java
@@ -96,7 +96,7 @@ public class QuizChecker {
                                 if (response != null) {
                                     setRevertParameter(response.getDeletedUploads());
                                 }
-                            })
+                            }, throwable -> Timber.e(throwable, "Fetching feedback failed"))
             );
     }
 

--- a/app/src/test/kotlin/fr/free/nrw/commons/mwapi/ApacheHttpClientMediaWikiApiTest.kt
+++ b/app/src/test/kotlin/fr/free/nrw/commons/mwapi/ApacheHttpClientMediaWikiApiTest.kt
@@ -6,6 +6,7 @@ import android.preference.PreferenceManager
 import com.google.gson.Gson
 import fr.free.nrw.commons.BuildConfig
 import fr.free.nrw.commons.TestCommonsApplication
+import okhttp3.OkHttpClient
 import okhttp3.mockwebserver.MockResponse
 import okhttp3.mockwebserver.MockWebServer
 import okhttp3.mockwebserver.RecordedRequest
@@ -30,14 +31,16 @@ class ApacheHttpClientMediaWikiApiTest {
     private lateinit var wikidataServer: MockWebServer
     private lateinit var sharedPreferences: SharedPreferences
     private lateinit var categoryPreferences: SharedPreferences
+    private lateinit var okHttpClient: OkHttpClient
 
     @Before
     fun setUp() {
         server = MockWebServer()
         wikidataServer = MockWebServer()
+        okHttpClient = OkHttpClient()
         sharedPreferences = PreferenceManager.getDefaultSharedPreferences(RuntimeEnvironment.application)
         categoryPreferences = PreferenceManager.getDefaultSharedPreferences(RuntimeEnvironment.application)
-        testObject = ApacheHttpClientMediaWikiApi(RuntimeEnvironment.application, "http://" + server.hostName + ":" + server.port + "/", "http://" + wikidataServer.hostName + ":" + wikidataServer.port + "/", sharedPreferences, categoryPreferences, Gson())
+        testObject = ApacheHttpClientMediaWikiApi(RuntimeEnvironment.application, "http://" + server.hostName + ":" + server.port + "/", "http://" + wikidataServer.hostName + ":" + wikidataServer.port + "/", sharedPreferences, categoryPreferences, Gson(), okHttpClient)
         testObject.setWikiMediaToolforgeUrl("http://" + server.hostName + ":" + server.port + "/")
     }
 

